dance/pipeline.py

```diff
@@ -848,15 +853,90 @@ def flatten_dict(d, *, parent_key="", sep="_"):
+    optional_elements.sort(key=lambda x: elements.index(x))
+
+    # Generate all possible combinations of optional elements
+    optional_combinations = []
+    for i in range(1, len(optional_elements) + 1):
+        optional_combinations += list(itertools.combinations(optional_elements, i))
+
+    # Combine required elements with optional combinations to get all possible combinations
+    all_combinations = []
+    for optional_combination in optional_combinations:
+        all_combinations.append([x for x in elements if x in required_elements or x in optional_combination])
+    return all_combinations
+
+
+def generate_subsets(path, tune_mode, save_directory, file_path, log_dir, required_indexes, save_config=True,
+                     root_path=None):
+    """Generated automatically subsets for original yaml. YAML can be generated
+    automatically, but obviously still need to manually tune the parameters to avoid
+    errors. When part of the process is omitted, the function parameters need to be
+    changed, otherwise an error will be reported, so different YAML adjustments are
+    required.
+
+    Conveniently generate multiple yamls, such as generating (A, B, C), (A, B), (A, C), (B, C), (A), (B), (C) yaml based on pipelines A, B, and C for the next experiment
+
+    Parameters
+    ----------
+    path
+        origin yaml file path
+    tune_mode
+        tune mode
+    save_directory
+        subset yaml files save path
+    file_path
+        python execution file, usually the file name is main
+    log_dir
+        log storage path
+    required_indexes
+        index in required type of origin yaml
+    Returns
+    --------
+    command_str: str
+        command_str,such as "python examples/tuning/cta_svm/main.py --config_dir=config_yamls/pipeline/subset_0_  --count=4 > temp_data/1.log 2>&1 &"
+    configs: list
+        yamls for pipeline subsets
+
+    """
+    DANCEDIR = Path(__file__).resolve().parent.parent
+    root_path = DANCEDIR if root_path is None else root_path
+    save_directory = f"{root_path}/{save_directory}"
+    path = f"{root_path}/{path}"
+    config = OmegaConf.load(path)
+    dict_config = DictConfig(config)
+    nums = dict_config[tune_mode]
+    subsets = generate_combinations_with_required_elements(nums, required_indexes)
+    configs = []
+    command_str = "#!/bin/bash \nlog_dir=" + log_dir + " \nmkdir -p ${log_dir}\n"
+    for index, subset in enumerate(subsets):
+        config_copy = dict_config.copy()
+        config_copy[tune_mode] = subset
+        configs.append(config_copy)
+        save_path = f"{save_directory}/subset_{index}_{tune_mode}_tuning_config.yaml"
+        if save_config:
+            OmegaConf.save(config_copy, save_path)
+        count = reduce(mul, [len(p["include"]) if "include" in p else 1 for p in subset])
+        config_dir = os.path.relpath(os.path.dirname(save_path), os.path.dirname(os.path.join(root_path, file_path)))
```

Comment:
Get the relative path between the file execution path and the configuration file path